src/stores/signaling.ts

```diff
@@ -0,0 +1,343 @@
+			const attendeeUsers = store.getters.participantsList(token) as Participant[]
+			const attendeeUsersToUpdate: InternalUpdatePayload = {}
+			const newSessions = new Set<string>()
+			let hasUnknownSessions = false
+
+			for (const participant of participants) {
+				newSessions.add(participant.sessionId)
+				const session = this.findSignalingSession(token, 'int-change', participant)
+				const attendeeId = session?.attendeeId
+				if (!attendeeId) {
+					hasUnknownSessions = true
+					continue
+				}
+
+				if (!attendeeUsersToUpdate[attendeeId]) {
+					// Prepare updated data
+					attendeeUsersToUpdate[attendeeId] = {
+						inCall: participant.inCall,
+						lastPing: participant.lastPing,
+						permissions: participant.participantPermissions,
+						sessionIds: [participant.sessionId],
+					}
+				} else {
+					// Participant might join from several devices
+					attendeeUsersToUpdate[attendeeId].sessionIds.push(participant.sessionId)
+				}
+			}
+
+			// Update participant objects
+			for (const attendee of attendeeUsers) {
+				const { attendeeId, sessionIds } = attendee
+				if (attendeeUsersToUpdate[attendeeId]) {
+					store.commit('updateParticipant', {
+						token,
+						attendeeId,
+						updatedData: attendeeUsersToUpdate[attendeeId],
+					})
+				} else if (sessionIds.length) {
+					// Participant left conversation from all devices
+					store.commit('updateParticipant', {
+						token,
+						attendeeId,
+						updatedData: { inCall: PARTICIPANT.CALL_FLAG.DISCONNECTED, sessionIds: [] },
+					})
+				}
+			}
+
+			// Clean up old sessions
+			for (const session of Object.keys(this.sessions)) {
+				if (!newSessions.has(session)) {
+					this.deleteSignalingSession(session)
+				}
+			}
+
+			return hasUnknownSessions
+		},
+
+		/**
+		 * Update participants joined in store according to data from standalone signaling server
+		 *
+		 * @param token the conversation token;
+		 * @param participants the newly joined participant objects;
+		 * @return {boolean} whether list has unknown sessions mapped to attendees list
+		 */
+		updateParticipantsJoinedFromStandaloneSignaling(token: string, participants: StandaloneSignalingJoinPayload[]): boolean {
+			const attendeeUsers = store.getters.participantsList(token) as Participant[]
+			const attendeeUsersToUpdate: Record<string, { displayName?: string, sessionIds: string[] }> = {}
+			let hasUnknownSessions = false
+
+			for (const participant of participants) {
+				const session = this.findSignalingSession(token, 'ext-join', participant)
+				const attendeeId = session?.attendeeId
+
+				if (!attendeeId) {
+					hasUnknownSessions = true
+					continue
+				}
+
+				const attendee = store.getters.getParticipant(token, attendeeId)
+
+				if (!attendeeUsersToUpdate[attendeeId]) {
+					attendeeUsersToUpdate[attendeeId] = { sessionIds: [...attendee.sessionIds] }
+				}
+				if (participant.user.displayname) {
```

Comment:
```suggestion
				if (participant.user?.displayname) {
```
index.js:2 Uncaught (in promise) TypeError: Cannot read properties of undefined (reading 'displayname')